docs/api/memory/link_time_optimization.md

```diff
@@ -0,0 +1,66 @@
+
+LTO profile build results for [mbed-cloud-client-example](https://github.com/ARMmbed/mbed-cloud-client-example)
+
+|Build type|Total Static RAM memory (data + BSS)|Total Flash memory (text + data)|
+| --- | --- | --- |
+| GCC_ARM - release - no LTO | 59,760B | 389,637B |
+| GCC_ARM - release - LTO | 59,432B | 354,167B |
+|***saved memory***| 328B | ‭35,470‬B|
+| ARM - release - no LTO | 58,099B | 353,849B |
+| ARM - release - LTO | 57,150B | 322,500B |
+|***saved memory***| 949B | ‭31,349‬B|
+
+<span class="notes">**Note**: In LTO builds, the compiler produces bytecode/bitcode instead of regular object code. It's hard to analyze this output with object code analysis tools.</span>
+
+## Limitations
+
+- LTO slows down build process.
+- It’s very hard to control memory placement when using LTO.
+- LTO performs heavy memory optimizations that break debugging.
+- In LTO builds, the compiler produce bytecode/bitcode instead of regular object code. It's hard to analyze this output with object code analysis tools.
+- LTO could cause increases to the stack space needed due to cross-object inlining.
+
+### Arm
+
+- No bitcode libraries: armlink only supports bitcode objects on the command line. It does not support bitcode objects coming from libraries. armlink gives an error message if it encounters a file containing bitcode while loading from a library.
+- Partial Linking is not supported with LTO as it only works with elf objects not bitcode files.
+- Arm recommends that link time optimization is only performed on code and data that does not require precise placement in the scatter file, with general input section selectors such as *(+RO) and .ANY(+RO) used to select sections generated by link time optimization. It is not possible to match bitcode in .llvmbc sections by name in a scatter file.
+- Bitcode objects are not guaranteed to be compatible across compiler versions. This means that you should ensure all your bitcode files are built using the same version of the compiler when linking with LTO.
+
+### IAR
+
+- There is no LTO available for the IAR compiler.
+
+### GCC_ARM
+
+- The minimal required version of `GCC_ARM` is now the GNU Arm Embedded Toolchain Version 9-2019-q4-major. Earlier `GCC_ARM` versions can cause various issues when the `-flto` flag is used, for example a platform-specific error during the final link stage on Windows hosts with GCC8.
+- The `noinline` attribute has to be used for every function that must be placed into a specific section (specified with a `section(".section_name")` attribute). In general, when a function is considered for inlining, the `section` attribute is always ignored. However, with the link-time optimizer enabled, the chances for inlining are much higher because the inliner works across multiple translation units. As a result, the output sections' sizes change compared to a non-lto build. This may lead to `section ".section_name" will not fit in region "region_name"` type errors.
```

Comment:
```suggestion
- You must use the `noinline` attribute for every function that must be placed into a specific section (specified with a `section(".section_name")` attribute). In general, when a function is considered for inlining, the `section` attribute is always ignored. However, with the link time optimizer enabled, the chances for inlining are much higher because the inliner works across multiple translation units. As a result, the output sections' sizes change compared to a non-lto build. This may lead to `section ".section_name" will not fit in region "region_name"` type errors.
```